controller/k8s/injector.go

```diff
@@ -0,0 +1,363 @@
+		fmt.Println("---")
+	}
+	return nil
+}
+
+/* Given a byte slice representing a deployment, unmarshal the deployment and
+ * return a new deployment with the sidecar and init-container injected.
+ */
+func injectDeployment(bytes []byte, podConfig PodConfig) (interface{}, error) {
+	var deployment v1beta1.Deployment
+	err := yaml.Unmarshal(bytes, &deployment)
+	if err != nil {
+		return nil, err
+	}
+	podTemplateSpec := podConfig.injectPodTemplateSpec(&deployment.Spec.Template)
+	return enhancedDeployment{
+		&deployment,
+		enhancedDeploymentSpec{
+			&deployment.Spec,
+			podTemplateSpec,
+		},
+	}, nil
+}
+
+/* Given a byte slice representing a replication controller, unmarshal the
+ * replication controller and return a new replication controller with the
+ * sidecar and init-container injected.
+ */
+func injectReplicationController(bytes []byte, podConfig PodConfig) (interface{}, error) {
+	var rc v1.ReplicationController
+	err := yaml.Unmarshal(bytes, &rc)
+	if err != nil {
+		return nil, err
+	}
+	podTemplateSpec := podConfig.injectPodTemplateSpec(rc.Spec.Template)
+	return enhancedReplicationController{
+		&rc,
+		enhancedReplicationControllerSpec{
+			&rc.Spec,
+			podTemplateSpec,
+		},
+	}, nil
+}
+
+/* Given a byte slice representing a replica set, unmarshal the replica set and
+ * return a new replica set with the sidecar and init-container injected.
+ */
+func injectReplicaSet(bytes []byte, podConfig PodConfig) (interface{}, error) {
+	var rs v1beta1.ReplicaSet
+	err := yaml.Unmarshal(bytes, &rs)
+	if err != nil {
+		return nil, err
+	}
+	podTemplateSpec := podConfig.injectPodTemplateSpec(&rs.Spec.Template)
+	return enhancedReplicaSet{
+		&rs,
+		enhancedReplicaSetSpec{
+			&rs.Spec,
+			podTemplateSpec,
+		},
+	}, nil
+}
+
+/* Given a byte slice representing a job, unmarshal the job and return a new job
+ * with the sidecar and init-container injected.
+ */
+func injectJob(bytes []byte, podConfig PodConfig) (interface{}, error) {
+	var job batchV1.Job
+	err := yaml.Unmarshal(bytes, &job)
+	if err != nil {
+		return nil, err
+	}
+	podTemplateSpec := podConfig.injectPodTemplateSpec(&job.Spec.Template)
+	return enhancedJob{
+		&job,
+		enhancedJobSpec{
+			&job.Spec,
+			podTemplateSpec,
+		},
+	}, nil
+}
+
+/* Given a byte slice representing a daemonset, unmarshal the daemonset and
+ * return a new daemonset with the sidecar and init-container injected.
+ */
+func injectDaemonSet(bytes []byte, podConfig PodConfig) (interface{}, error) {
+	var ds v1beta1.DaemonSet
+	err := yaml.Unmarshal(bytes, &ds)
+	if err != nil {
+		return nil, err
+	}
+	podTemplateSpec := podConfig.injectPodTemplateSpec(&ds.Spec.Template)
+	return enhancedDaemonSet{
+		&ds,
+		enhancedDaemonSetSpec{
+			&ds.Spec,
+			podTemplateSpec,
+		},
+	}, nil
+}
+
+/* Given a PodTemplateSpec, return a new PodTemplateSpec with the sidecar
+ * and init-container injected.
+ */
+func (p PodConfig) injectPodTemplateSpec(t *v1.PodTemplateSpec) enhancedPodTemplateSpec {
```

Comment:
i think how `injectPodTemplateSpec` is a method off of `PodConfig`. any reason not to give the same treatment to the other `inject*` functions?

Reply:
> i think how injectPodTemplateSpec is a method off of PodConfig. any reason not to give the same treatment to the other inject* functions?

What's the normal pattern in our code, in k8s client-go, in Kubernetes in general, and in Go code in general? It seems the norm is for config objects to be passed as parameters to standalone functions, from what I remember.

Reply:
Reverting the `injectYAML` will no longer require us to have a `PodConfig`